README.md
# WebUSB NeoPixel Controller

[![1.0.0](https://img.shields.io/badge/Stable-1.0.0-00f557.svg?style=flat)](https://github.com/NERDDISCO/webusb-neopixel-controller/releases/tag/1.0.0)

Manage a NeoPixels with an Arduino-based controller directly from the browser by leveraging WebUSB.

👉 To fully understand the background of this module you should read the article [How to build a WebUSB DMX512 Controller by using an Arduino](https://medium.com/@timpietrusky/how-to-build-a-webusb-dmx512-controller-by-using-an-arduino-e0dd8efb7bf0). 👈

## Table of Contents

<!-- toc -->

- [ES6 module](#es6-module)
  * [Install](#install)
  * [API Documentation](#api-documentation)
  * [Usage](#usage)
  * [Browser Support](#browser-support)
- [Arduino](#arduino)
  * [Models](#models)
  * [Sketch](#sketch)
- [NeoPixels](#neopixels)
  * [Connect to Arduino](#connect-to-arduino)
- [Demo](#demo)
  * [Run locally](#run-locally)

<!-- tocstop -->

---

## ES6 module

The module can be used in projects where you want to control a NeoPixel controller over WebUSB. It's written in JavaScript and contains only one Class called `Controller`.

### Install

Install the `webusb-neopixel-controller` module into your project:

```
npm install webusb-neopixel-controller
```


### API Documentation

Can be found on [API documentation](docs/API.md).


### Usage

```javascript
import Controller from 'webusb-neopixel-controller/controller.js'

const controller = new Controller({ leds: 30 })
const activateButton = document.getElementById('activateWebUsb')

// Listen for click events on the activate button, because
// `controller.enable` must be triggered by a user gesture
activateButton.addEventListener('click', e => {

  // Enable WebUSB and select the Arduino
  controller.enable().then(() => {

    // Create a connection to the selected Arduino
    controller.connect().then(() => {

      // Set the 1 LED of the NeoPixel strip to yellow (rgb(255, 255, 0))
      controller.update(1, [255, 0, 255])
    })
  })
})
```

Also make sure to take a look the [code behind the demo](#demo) to get more usage examples.

The module is also used in [luminave](https://github.com/NERDDISCO/luminave).


### Browser Support

In order to use the module you have to use a browser that supports WebUSB:

* Chrome 61 + 62: Behind flags
  * `chrome://flags/#enable-experimental-web-platform-features`
* Chrome 63+: Native support

---

## Arduino 

To transform the Arduino into a WebUSB NeoPixel Controller, you need to get some code:


```bash
# Clone the repository
git clone git@github.com/NERDDISCO/webusb-neopixel-controller.git

# Get dependencies
git submodule init
git submodule update
```

It clones the following submodules:

* [Adafruit_NeoPixel](https://github.com/adafruit/Adafruit_NeoPixel)

### Models

The code was tested on these Arduino:

* [Arduino Leonardo](https://store.arduino.cc/arduino-leonardo-with-headers)
* Arduino Leonardo ETH
* Seeeduino Lite


### Sketch

The code that runs on the Arduino is organized in a Sketch. In order to use the sketch from this repository, you have to follow these steps:

1. Download & install the [Arduino IDE](https://www.arduino.cc/en/Main/Software#download) >= 1.8.5, so you are able to write & push code onto the Arduino
2. Open the Arduino IDE
3. Open the preferences: *Arduino > Preferences*
4. In the preferences dialog you have to change the *Sketchbook location* so that it points to the *sketchbook* folder that comes with the repository:

  ![Arduino IDE: Change sketchbook location](docs/images/arduino_ide_preferences_sketchbook_location.png)
5. Close the Arduino IDE and then open it again (this is needed to load the new sketchbook that we selected in the step before)
6. Now we need to configure the Arduino IDE so that it can recognize our Arduino Leonardo:
   1. Select the model: *Tools > Board > Arduino Leonardo (WebUSB)* (or one of the other supported boards underneath the *NERDDISCO: WebUSB NeoPixel Controller* headline in the menu)
     ![Arduino IDE: Select the Board](docs/images/arduino_ide_select_board.png)
   2. Select the USB port: *Tools > Port > /dev/tty.usbmodem$* (The $ is a placeholder and will be different for every Arduino you are using)

    **Attention**: This can only be selected if your Arduino is actually attached to your computer via USB!
7. Open the sketch (if it's not already open): *File > Sketchbook > webusb_neopixel_controller*
8. Verify that the sketch is working: *Sketch > Verify/Compile*. This will produce an output like this:
```sh
  Sketch uses 7624 bytes (26%) of program storage space. Maximum is 28672 bytes.
  Global variables use 569 bytes (22%) of dynamic memory, leaving 1991 bytes for local variables. Maximum is 2560 bytes.
```
9. Upload the sketch to the Arduino: *Sketch > Upload* This will produce a similar output as step 10.

**Attention**: If the steps 8 or 9 did not work as promised, [please open an issue on GitHub](https://github.com/NERDDISCO/webusb-neopixel-controller/issues)

When you are done your Arduino IDE should look like this:

![Arduino IDE: Full verified sketch](docs/images/arduino_ide_sketch_verified.png)


---

## NeoPixels

You can use any [NeoPixels](https://www.adafruit.com/category/168) and (in theory) every other LED that is based on the WS2812 or SK6812 chipset. 

### Connect to Arduino

For testing purposes you can connect the NeoPixels directly to the Arduino. 

**Caution**: This only works if you are using less than 30 NeoPixels at the same time. Everything else will require an external power source. In order to handle that you [should read the guide from Adafruit](https://learn.adafruit.com/adafruit-neopixel-uberguide/basic-connections). 

**Caution**: The order of the three pins can vary between different strip densities and batches. ALWAYS use the labels printed ON THE STRIP. Look closely, NEVER blindly follow a NeoPixel strip wiring diagram; it might be based on a different strip type!

![NeoPixel connected to Arduino](docs/images/arduino_webusb_neopixel_fritzing.jpg)


---


## Demo

In order to test the WebUSB NeoPixel Controller directly in the browser you can use the [demo on GitHub](https://nerddisco.github.io/webusb-neopixel-controller). The [code behind the demo](https://github.com/NERDDISCO/webusb-neopixel-controller/tree/master/demo) can be found in the repository.

### Run locally

* Clone [this repository](https://github.com/NERDDISCO/webusb-neopixel-controller)
* Install the dev dependencies by executing `npm install` inside the repository
* Execute `npm start` to start the local web server
* Open the demo on [localhost:8080](http://localhost:8080)

docs/CHANGELOG.md
<a name="1.0.1"></a>
## [1.0.1](https://github.com/NERDDISCO/webusb-neopixel-controller/compare/v1.0.0...v1.0.1) (2018-11-27)


### Bug Fixes

* **build:** Package should be public on npm ([7981265](https://github.com/NERDDISCO/webusb-neopixel-controller/commit/7981265))

<a name="1.0.0"></a>
# 1.0.0 (2018-11-27)


### Features

* **hello world:** yes, neopixels too! ([33fb1fe](https://github.com/NERDDISCO/webusb-neopixel-controller/commit/33fb1fe))


tests/controller.test.js
import Controller from '../controller.js'

test('Use controller with default arguments', () => {
  const controller = new Controller({ leds: 30 })

  // Device not defined
  expect(controller.device).toBe(undefined)

  // leds is an array with 30 leds
  expect(Array.isArray(controller.leds)).toBe(true)
  expect(controller.leds.length).toBe(30 * 3)

  // Filters is an array with 7 filtered devices
  expect(Array.isArray(controller.filters)).toBe(true)
  expect(controller.filters.length).toBe(7)
})



test('Use controller with custom arguments (device, leds and filters)', () => {
  const controller = new Controller({
    filters: [],
    leds: 10,
    device: {}
  })

  expect(controller.filters).toEqual([])
  expect(controller.leds.length).toEqual(10 * 3)
  expect(controller.device).toEqual({})
})



test('Arduino Leonardo is allowed', () => {
  const controller = new Controller({ leds: 30 })

  expect(controller.filters).toContainEqual(
    expect.objectContaining({ vendorId: 0x2341, productId: 0x8036 })
  )
  expect(controller.filters).toContainEqual(
    expect.objectContaining({ vendorId: 0x2341, productId: 0x0036 })
  )
  expect(controller.filters).toContainEqual(
    expect.objectContaining({ vendorId: 0x2a03, productId: 0x8036 })
  )
  expect(controller.filters).toContainEqual(
    expect.objectContaining({ vendorId: 0x2a03, productId: 0x0036 })
  )
})



test('Arduino Leonardo ETH is allowed', () => {
  const controller = new Controller({ leds: 30 })

  expect(controller.filters).toContainEqual(
    expect.objectContaining({ vendorId: 0x2a03, productId: 0x0040 })
  )
  expect(controller.filters).toContainEqual(
    expect.objectContaining({ vendorId: 0x2a03, productId: 0x8040 })
  )
})



test('Seeeduino Lite is allowed', () => {
  const controller = new Controller({ leds: 30 })

  expect(controller.filters).toContainEqual(
    expect.objectContaining({ vendorId: 0x2886, productId: 0x8002 })
  )
})


demo/index.js
import Controller from '../controller.js'
import DevConsole from './dev-console.js'

import { HLSToRGB } from 'hsl-to-rgb-lightweight'

const leds = 60
const ledBytes = leds * 3
const delay = 1

const controller = window.controller = new Controller({ leds: new Array(ledBytes).fill(0) })
const devConsole = new DevConsole({
  output : document.getElementById('console'),
  logLedsElement : document.getElementById('logLeds')
})

const activateButton = document.getElementById('activateWebUsb')
const disconnectButton = document.getElementById('disconnectWebUsb')

const setChannelForm = document.getElementById('updateAnyChannel')

const playBlack = document.getElementById('animationBlack')
const playRainbow = document.getElementById('animationRainbow')
let isAnimationPlaying = false


// Listen for click events on the activate button
activateButton.addEventListener('click', e => {

  // Enable WebUSB and select the Arduino
  controller.enable().then(() => {

    // Create a connection to the selected Arduino
    controller.connect().then(() => {

      // Successfully created a connection to the device
      devConsole.logUsbDevice(controller.device)
    })
  })
  .catch(() => {
    devConsole.log('No USB device was selected', '', 'string')
  })

})



// Disconnect from USB device
disconnectButton.addEventListener('click', e => {
  controller.disconnect().then(() => {
    devConsole.log('Destroyed connection to USB device, but USB device is still paired with the browser', '', 'string')
  })
})



// Automatically connect to paired USB device
controller.autoConnect()
  .then(() => {
    devConsole.log('Found an already paired USB device', '', 'string')
    devConsole.logUsbDevice(controller.device)
  })
  .catch((error) => {
    devConsole.log('autoConnect:', error, 'string')
  })



// Listen for submit events to change the value of a channel
setChannelForm.addEventListener('submit', e => {
  e.preventDefault()

  // Get data out of form
  const data = new FormData(setChannelForm)

  // Parse the data into an Integer
  const channel = parseInt(data.get('channel'), 10)
  const color = data.get('color').match(/[A-Za-z0-9]{2}/g).map(v => parseInt(v, 16))

  devConsole.log('---', '', 'string')
  devConsole.log(`Set LED ${channel} to ${color}`, '', 'string')

  // Update the leds
  controller.update(channel, color)
  .then(() => {
    devConsole.logLeds(controller.leds)
  })
  .catch((error) => {
    devConsole.logLeds(controller.leds)
    devConsole.log(error, '', 'string')
  })
})

// Listen for click events on the black animation button
playBlack.addEventListener('click', e => {
  controller.leds.splice(0, controller.leds.length, ...new Array(ledBytes).fill(0))

  devConsole.log('Set all NeoPixel to black', '', 'string')

  controller.send(controller.leds)
  .then(() => {
    devConsole.logLeds(controller.leds)
  })
  .catch(error => {
    devConsole.logLeds(controller.leds)
    devConsole.log(error, '', 'string')
  })
})

// Listen for click events on the rainbow animation button
playRainbow.addEventListener('click', e => {
  isAnimationPlaying = !isAnimationPlaying
  playRainbow.innerHTML = playRainbow.innerHTML === 'Play' ? 'Pause' : 'Play'

  if (isAnimationPlaying) {
    devConsole.log('Started Rainbow animation', '', 'string')

    loop()
  } else {
    devConsole.log('Stopped Rainbow animation', '', 'string')
  }
})


let colorIndex = 0

const loop = () => {
  if (isAnimationPlaying) {

    for (let i = 0; i < leds; i++) {
      let rgb = HLSToRGB.convert((colorIndex + (i * delay)) % 360, 1, 0.4)
      controller.leds.splice(i * 3, rgb.length, ...rgb)
    }

    colorIndex = ++colorIndex > 360 ? 0 : colorIndex
      
    // Send the updated leds to the controller
    controller.send(controller.leds)
    .then(() => {
    })
    .catch(error => {
      devConsole.logLeds(controller.leds)
      devConsole.log(error, '', 'string')
    })

    window.requestAnimationFrame(loop)
  }
}

demo/dev-console.js
/**
 * Helper to log messages into an <code>output</code> DOM element and the
 * console in the browser.
 */
export default class DevConsole {
  constructor(args = {}) {
    // The element to write the log messages to
    this.output = args.output

    // The element to toggle "logLedsEnabled"
    this.logLedsElement = args.logLedsElement

    // Is logging the content of the universe enabled?
    this.logLedsEnabled = false

    // Listen for changes to enable / disable "log universe to console"
    this.logLedsElement.addEventListener('change', e => {
      const { target } = e

      this.logLedsEnabled = target.checked
    })
  }

  /**
   * Log a message to the <code>output</code> and the browser console
   */
  log(message, data, type) {
    let fullMessage = ''

    switch (type) {
      case 'USBDevice':
        fullMessage = `${message}: ${data}`
        break

      case 'array':
        fullMessage = message + JSON.stringify(data)
        break

      case 'keyvalue':
        fullMessage = `${message}: ${data}`
        break

      default:
        fullMessage = message + ' ' + data
    }

    console.log(fullMessage)

    this.output.value += fullMessage + '\n'

    // Automatically scroll to the bottom
    this.output.scrollTop = this.output.scrollHeight
  }

  /*
   * Basic information about the USB device
   */
   logUsbDevice(device) {
    this.log('---', '', 'string')
    this.log('Selected device', device.productName, 'USBDevice')
    this.log('---', '', 'string')

    const { configuration, configurations, deviceClass, deviceProtocol,
      deviceSubclass, deviceVersionMajor, deviceVersionMinor,
      deviceVersionSubminor, manufacturerName, opened, productId,
      productName, serialNumber, usbVersionMajor, usbVersionMinor,
      usbVersionSubminor, vendorId } = device

    // Convert IDs from decimal to hex
    const vendorIdHex = vendorId.toString(16)
    const productIdHex = productId.toString(16)

    this.log('Opened', opened, 'keyvalue')
    this.log('Vendor ID', `${vendorId} (0x${vendorIdHex})`, 'keyvalue')
    this.log('Manufacturer Name', manufacturerName, 'keyvalue')
    this.log('Product ID', `${productId} (0x${productIdHex})`, 'keyvalue')
    this.log('Product Name', productName, 'keyvalue')
    this.log('Serialnumber', serialNumber, 'keyvalue')

    this.log('Device Class', deviceClass, 'keyvalue')
    this.log('Device Protocol', deviceProtocol, 'keyvalue')
    this.log('Device Subclass', deviceSubclass, 'keyvalue')
    this.log('Device Version Major', deviceVersionMajor, 'keyvalue')
    this.log('Device Version Minor', deviceVersionMinor, 'keyvalue')
    this.log('Device Version Subminor', deviceVersionSubminor, 'keyvalue')

    this.log('USB Version Major', usbVersionMajor, 'keyvalue')
    this.log('USB Version Minor', usbVersionMinor, 'keyvalue')
    this.log('USB Version Subminor', usbVersionSubminor, 'keyvalue')
  }

  /**
   * The whole content of the universe if <code>logLedsEnabled</code>
   */
  logLeds(universe) {
    if (this.logLedsEnabled) {
      this.log('', universe, 'array')
    }
  }
}


controller.js
/**
 * The controller is creating a connection to the USB device (Arduino) to send data over WebUSB.
 * By using the default <code>args</code> you will only see the following Arduino in the user prompt:
 * - Arduino Leonardo
 * - Arduino Leonardo ETH
 * - Seeeduino Lite
 * @module Controller
 *
 * @param {Object} args - Arguments to configure the controller
 * @param {Object[]} args.filters - List of devices that are whitelisted when opening the user prompt to select an Arduino
 * @param {Object} args.device - The selected Arduino to use as the controller
 * @param {number[]} args.leds - Holds all the values for each led of the leds
 * @example
 * import Controller from 'webusb-neopixel-controller/controller.js'
 *
 * // Create a new controller using the default properties
 * const controller = new Controller()
 */
export default class Controller {

  constructor(args = {}) {
    // Reference to the selected USB device
    this.device = args.device || undefined

    // Only allow specific USB devices
    this.filters = args.filters || [
      // Arduino Leonardo
      { vendorId: 0x2341, productId: 0x8036 },
      { vendorId: 0x2341, productId: 0x0036 },
      { vendorId: 0x2a03, productId: 0x8036 },
      { vendorId: 0x2a03, productId: 0x0036 },

      // Arduino Leonardo ETH
      { vendorId: 0x2a03, productId: 0x0040 },
      { vendorId: 0x2a03, productId: 0x8040 },

      // Seeeduino Lite
      { vendorId: 0x2886, productId: 0x8002 }
    ]

    // Initialize the array that holds the NeoPixels
    this.leds = new Array(args.leds * 3).fill(0)
  }

  /**
   * Enable WebUSB and save the selected Arduino into <code>controller.device</code>
   *
   * Note: This function has to be triggered by a user gesture
   *
   * @return {Promise}
   *
   * @example
   * controller.enable().then(() => {
   *   // Create a connection to the selected Arduino
   *   controller.connect().then(() => {
   *     // Successfully created a connection
   *   })
   * })
   * .catch(error => {
   *   // No Arduino was selected by the user
   * })
   */
  enable() {
    // Request access to the USB device
    return navigator.usb.requestDevice({ filters: this.filters })

    // selectedDevice = the USB device that was selected by the user in the browser
    .then(selectedDevice => {
      this.device = selectedDevice
    })
  }

  /**
   * Get a USB device that was already paired with the browser.
   *
   * @return {Promise}
   */
  getPairedDevice() {
    return navigator.usb.getDevices()

    .then(devices => {
      return devices[0]
    })
  }

  /**
   * Automatically connect to a USB device that was already paired with the Browser and save it into <code>controller.device</code>
   *
   * @return {Promise}
   * @example
   * controller.autoConnect()
   *   .then(() => {
   *     // Connected to already paired Arduino
   *   })
   *   .catch(error => {
   *     // Nothing found or found paired Arduino, but it's not connected to computer
   *   })
   */
  autoConnect() {
    return this.getPairedDevice().then((device) => {

      this.device = device

      return new Promise((resolve, reject) => {

        // USB Device is not connected to the computer
        if (this.device === undefined) {
          return reject(new Error('Can not find USB device.'))

        // USB device is connected to the computer, so try to create a WebUSB connection
        } else {
          return resolve(this.connect())
        }

      })

    })
  }

  /**
   * Open a connection to the selected USB device and tell the device that
   * we are ready to send data to it.
   *
   * @return {Promise}
   * @example
   * controller.connect().then(() => {
   *   // Successfully created a connection to the selected Arduino
   * })
   */
  connect() {
    // Open connection
    return this.device.open()

    // Select #1 configuration if not automatially set by OS
    .then(() => {
      if (this.device.configuration === null) {
        return this.device.selectConfiguration(1)
      }
    })

    // Get exclusive access to the #2 interface
    .then(() => this.device.claimInterface(2))

    // Tell the USB device that we are ready to send data
    .then(() => this.device.controlTransferOut({
        // It's a USB class request
        'requestType': 'class',
        // The destination of this request is the interface
        'recipient': 'interface',
        // CDC: Communication Device Class
        // 0x22: SET_CONTROL_LINE_STATE
        // RS-232 signal used to tell the USB device that the computer is now present.
        'request': 0x22,
        // Yes
        'value': 0x01,
        // Interface #2
        'index': 0x02
      })
    )

    .then(() => {
      this.readLoop()
    })
  }

  /**
   * Read data from the NeoPixel Controller
   *
   * @return {Promise}
   */
  readLoop() {
    // Get data from the USB device on Endpoint #5
    this.device.transferIn(5, 1).then(result => {
      var amountLeds = new Uint8Array(result.data.buffer)
      console.log(`Connected NeoPixels: ${amountLeds}`)

      this.readLoop()
    }, error => {
      console.log(error)
    })
  }

  /**
   * Send data to the USB device to update the leds
   *
   * @param {Array} data - List containing all leds that should be updated
   *
   * @return {Promise}
   * @example
   * controller.send([255, 0, 0])
   */
  send(data) {
    return new Promise((resolve, reject) => {

      // USB Device is not connected to the computer
      if (this.device === undefined) {
        return reject(new Error('USB device is not connected to the computer'))

      // USB device is connected to the computer, so try to create a WebUSB connection
      } else {
        // Create an ArrayBuffer, because that is needed for WebUSB
        const buffer = Uint8Array.from(data)

        // Send data on Endpoint #4
        return resolve(this.device.transferOut(4, buffer))
      }

    })
  }

  /**
   * Update the <code>led</code>(s) of the leds with the provided <code>value</code>
   *
   * @param {number} led - The led to update
   * @param {(number|number[])} value - The value to update the led, supporting two different modes: single (= <code>number</code>) & multi (= <code>Array</code>)
   * @example <caption>Update a single led</caption>
   * // Update led #1 with color red (rgb(255, 0, 0))
   * controller.update(1, [255, 0, 0])
   * @example <caption>Update multiple leds starting with led</caption>
   * // Update led #5 with red (rgb(255, 0, 0)), #6 with green (rgb(0, 255, 0))
   * controller.update(5, [255, 0, 0, 0, 255, 0])
   */
  update(led, value) {
    return new Promise((resolve, reject) => {

      // (led - 1): The LEDs start with led 1, but the array with 0
      // * 3: Every LED can display an RGB color, so every 3 bytes in the array are one LED
      led = (led - 1) * 3

      // Multiple
      if (Array.isArray(value)) {
        this.leds.splice(led, value.length, ...value)
      
      } else {
        return reject(new Error('Could not update LEDs because the provided value is not of type number[]'))
      }

      // Send the updated leds to the controller
      return resolve(this.send(this.leds))

    })
  }

  /**
   * Disconnect from the USB device
   *
   * Note: The device is still paired to the browser!
   *
   * @return {Promise}
   * @example
   * controller.disconnect().then(() => {
   *   // Destroyed connection to USB device, but USB device is still paired with the browser
   *})
   */
  disconnect() {
    // Declare that we don't want to receive data anymore
    return this.device.controlTransferOut({
      // It's a USB class request
      'requestType': 'class',
      // The destination of this request is the interface
      'recipient': 'interface',
      // CDC: Communication Device Class
      // 0x22: SET_CONTROL_LINE_STATE
      // RS-232 signal used to tell the USB device that the computer is not present anymore
      'request': 0x22,
      // No
      'value': 0x01,
      // Interface #2
      'index': 0x02
    })

    // Close the connection to the USB device
    .then(() => this.device.close())
  }
}
